File: gateway/enforcer/org.wso2.apk.enforcer/src/main/java/org/wso2/apk/enforcer/models/Subscription.java
/*
 * Copyright (c) 2020, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.wso2.apk.enforcer.models;

import org.wso2.apk.enforcer.common.CacheableEntity;

/**
 * Entity for representing a SubscriptionDTO in APIM.
 */
public class Subscription implements CacheableEntity<String> {

    private String subscriptionId = null;
    private String subscriptionStatus = null;
    private String organization = null;
    private SubscribedAPI subscribedApi = null;
    private String ratelimitTier = null;
    private long timeStamp;

    public String getSubscriptionId() {

        return subscriptionId;
    }

    public void setSubscriptionId(String subscriptionId) {

        this.subscriptionId = subscriptionId;
    }

    public String getSubscriptionStatus() {

        return subscriptionStatus;
    }

    public void setSubscriptionStatus(String subscriptionStatus) {

        this.subscriptionStatus = subscriptionStatus;
    }

    public String getOrganization() {

        return organization;
    }

    public void setOrganization(String organization) {

        this.organization = organization;
    }

    public SubscribedAPI getSubscribedApi() {

        return subscribedApi;
    }

    public void setSubscribedApi(SubscribedAPI subscribedApi) {

        this.subscribedApi = subscribedApi;
    }

    public long getTimeStamp() {

        return timeStamp;
    }

    public void setTimeStamp(long timeStamp) {

        this.timeStamp = timeStamp;
    }

    @Override
    public String getCacheKey() {

        return subscriptionId;
    }

    public String getRatelimitTier() {
        return ratelimitTier;
    }

    public void setRatelimitTier(String ratelimitTier) {
        this.ratelimitTier = ratelimitTier;
    }

    @Override
    public String toString() {

        return "Subscription{" +
                "subscriptionId='" + subscriptionId + '\'' +
                ", subscriptionStatus='" + subscriptionStatus + '\'' +
                ", organization='" + organization + '\'' +
                ", subscribedApi=" + subscribedApi +
                ", timeStamp=" + timeStamp +
                '}';
    }
}

File: test/integration/integration/tests/subscription.go
/*
 *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package tests

import (
	"testing"

	"github.com/wso2/apk/test/integration/integration/utils/http"
	"github.com/wso2/apk/test/integration/integration/utils/suite"
)

func init() {
	IntegrationTests = append(IntegrationTests, Subscription)
}

// Subscription test
var Subscription = suite.IntegrationTest{
	ShortName:   "Subscription",
	Description: "Tests the subscription functionality",
	Manifests:   []string{"tests/subscription.yaml"},
	Test: func(t *testing.T, suite *suite.IntegrationTestSuite) {
		ns := "gateway-integration-test-infra"
		gwAddr := "all-http-methods-for-wildcard.test.gw.wso2.com:9095"
		clientToken := http.GetClientCredentialsToken(t, "45f1c5c8-a92e-11ed-afa1-0242ac120065")
		appToken := http.GetAppClientCredentialsToken(t, "llm-app")
		token := http.GetTestToken(t)
		testCases := []http.ExpectedResponse{
			// test path with trailing slash for GET
			{
				TestCaseName: "test path with trailing slash for GET",
				Request: http.Request{
					Host:   "all-http-methods-for-wildcard.test.gw.wso2.com",
					Path:   "/all-http-methods-for-wildcard/v1.0.0/v2/echo-full/",
					Method: "GET",
				},
				ExpectedRequest: &http.ExpectedRequest{
					Request: http.Request{
						Path: "/v2/echo-full/",
					},
				},
				Backend:   "infra-backend-v1",
				Namespace: ns,
			},
		}

		testCasesWithNoSubsToken := []http.ExpectedResponse{
			// test path with trailing slash for GET
			{
				TestCaseName: "test path with trailing slash for GET with no subscription token",
				Request: http.Request{
					Host:   "all-http-methods-for-wildcard.test.gw.wso2.com",
					Path:   "/all-http-methods-for-wildcard/v1.0.0/v2/echo-full/",
					Method: "GET",
				},
				ExpectedRequest: &http.ExpectedRequest{
					Request: http.Request{
						Path: "/v2/echo-full/",
					},
				},
				Backend:   "infra-backend-v1",
				Namespace: ns,
				Response: http.Response{
					StatusCode: 403,
				},
			},
		}

		for i := range testCases {
			tc := testCases[i]
			tc.Request.Headers = http.AddBearerTokenToHeader(clientToken, tc.Request.Headers)
			t.Run(tc.GetTestCaseName(i), func(t *testing.T) {
				t.Parallel()
				http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, tc)
			})
		}
		for i := range testCases {
			tc := testCases[i]
			tc.Request.Headers = http.AddBearerTokenToHeader(appToken, tc.Request.Headers)
			t.Run(tc.GetTestCaseName(i), func(t *testing.T) {
				t.Parallel()
				http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, tc)
			})
		}
		for i := range testCasesWithNoSubsToken {
			tc := testCasesWithNoSubsToken[i]
			tc.Request.Headers = http.AddBearerTokenToHeader(token, tc.Request.Headers)
			t.Run(tc.GetTestCaseName(i), func(t *testing.T) {
				t.Parallel()
				http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, tc)
			})
		}
	},
}
